### PR TITLE
[WIP] Fix `logits/chosen` and `logits/rejected` metrics in `kto_trainer`.

### PR DESCRIPTION
**Pull request by PhilipMay**
The calculation of the `logits/chosen` and `logits/rejected` metrics in `kto_trainer` seem to be wrong. A `nansum()` followed by `nanmean()` applied to the `policy_rejected_logits` is wrong.

Our fix is to apply `nansum()` followed by an other `nansum()` and then devide the result by `count/chosen` or `count/rejected`.

### CODE REVIEW

**Review by kashif — APPROVED** [review]

**Comment by PhilipMay** [comment]
Tagging @MAOJIASONG as the original author and @claralp and asking for a review.

**Comment by qgallouedec** [comment]
Now I understand:

Let's take the following example:

```
              process 1   process 2
logits        [0, 1, 2]   [3, 4]
sum               3         7
gather              [3, 7]
mean                  5
```

Your proposed fix

```
              process 1   process 2
logits        [0, 1, 2]   [3, 4]
sum               3         7
gather              [3, 7]
sum                   10
/all_num_chosen        2
```

Would the following work?

```
              process 1   process 2
logits        [0, 1, 2]   [3, 4]
gather          [0, 1, 2, 3, 4]
sum                   10
/all_num_chosen        2
```

**Review by qgallouedec — APPROVED** [review]

**Comment by MAOJIASONG** [comment]
> Now I understand:
> 
> Let's take the following example:
> 
> ```
>               process 1   process 2
> logits        [0, 1, 2]   [3, 4]
> sum               3         7
> gather              [3, 7]
> mean                  5
> ```
> 
> Your proposed fix
> 
> ```
>               process 1   process 2
> logits        [0, 1, 2]   [3, 4]
> sum               3         7
> gather              [3, 7]
> sum                   10
> /all_num_chosen        2
> ```
> 
> Would the following work?
> 
> ```
>               process 1   process 2
> logits        [0, 1, 2]   [3, 4]
> gather          [0, 1, 2, 3, 4]
> sum                   10
> /all_num_chosen        2
> ```
> 
> Probably, but too memory intensive, right?

May I ask what's the difference between the first one, and the proposed second one? why is it better to use `all_num_chosen` for division? @qgallouedec

**Comment by qgallouedec** [comment]
Because otherwise the result is wrong (5)

**Comment by PhilipMay** [comment]
> Probably, but too memory intensive, right?

@qgallouedec I am not sure about this to be honest.

**Comment by MAOJIASONG** [comment]
> Because otherwise the result is wrong (5)

Thx, I misinterpreted the example. Thanks for point it out.
